### PR TITLE
Add stash entry type

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
Adds a type to represent a stash object.

See tracking issue: #6987

### CODE REVIEW

**Comment by iAmWillShepherd** [comment]
Note to whoever reviews this PR:
I'm looking for feedback on whether this is the right place for this type. I think a case can be made to put this in the `models` directory, but I threw it in this new file as a start.

**Comment by shiftkey** [comment]
> I think a case can be made to put this in the `models` directory, but I threw it in this new file as a start.

@iAmWillShepherd I'd love to understand how broadly you imagine this type being used, because depending on the answer I can see a couple of options:

 - if you can see this being returned by Git operations, stored in application state, and used in components, then you could make a case for this living under `app/src/models`
 - if you can see thing being returned by the Git operation, but being mapped to something else that is then used for application state and components, I'd keep this under `app/src/lib/git` for the moment

Note that this might change over time as you get the functionality in place, but for now I'd keep things as simple and localized as possible.

**Comment by iAmWillShepherd** [comment]
@outofambit, @shiftkey at the moment, this type will be used to represent a stash entry from git, in the future, I may use it in app-state, but it's a bit early for me to know for sure. I'm 👍 for keeping it in `lib/git` until it's needed in more than one place.

**Comment by shiftkey** [comment]
@iAmWillShepherd any other notes about what you'd like reviewed?

**Comment by iAmWillShepherd** [comment]
> @iAmWillShepherd any other notes about what you'd like reviewed?

Actually yes, I'm wondering if the `sha` is necessary.

My original plan for including the SHA was to handle scenarios where a branch is renamed or if we have an ambiguous branch name. My thinking was that we could use the SHA to verify if we are using the right branch.

**Comment by shiftkey** [comment]
> My thinking was that we could use the SHA to verify if we are using the right branch.

What does "right branch" mean in this context?

**Comment by iAmWillShepherd** [comment]
> > My thinking was that we could use the SHA to verify if we are using the right branch.
> 
> What does "right branch" mean in this context?

Right branch in this context means I have the correct stash entry for the current branch. After thinking on this some more, I think I'm going to use the SHA to refer to the stash entry instead of using it to store the tip of the branch.

File: app/src/lib/git/stash.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export interface IStashEntry {
+  /** The name of the branch at the time the entry was created. */
+  readonly branchName: string
+
+  /** The SHA of the commit object created as a result of stashing. */
+  readonly sha: string
```
Comment (shiftkey) [inline app/src/lib/git/stash.ts]: 💭 could this be something more descriptive than `sha`?

Comment (iAmWillShepherd) [reply app/src/lib/git/stash.ts]: Changed - https://github.com/desktop/desktop/pull/6948/commits/01b9731821e6da77b4e2bd3d3c873cd3c32fe819

**Comment by iAmWillShepherd** [comment]
🍎

**Comment by shiftkey** [comment]
@iAmWillShepherd feel free to merge whenever you're ready to move on

**Pull request merged**